src/cloudant/index.py

```diff
@@ -0,0 +1,173 @@
+#!/usr/bin/env python
+"""
+_utils_
+
+General utilities
+
+
+"""
+import json
+
+from collections import Sequence
+
+ALL_ARGS = [
+    "descending",
+    "endkey",
+    "endkey_docid",
+    "group",
+    "group_level",
+    "include_docs",
+    "inclusive_end",
+    "key",
+    "limit",
+    "reduce",
+    "skip",
+    "stale",
+    "startkey",
+    "startkey_docid"
+]
+
+BOOLEAN_ARGS = [
+    'include_docs',
+    'inclusive_end',
+    'reduce',
+    'group',
+    'descending'
+]
+
+ARRAY_ARGS = [
+    'startkey',
+    'endkey'
+]
+STRING_ARGS = [
+    'key',
+]
+
+
+def python_to_couch(kwargs):
+    """
+    _python_to_couch_
+
+    Translator method to flip python style
+    options into couch query options, eg True => 'true'
+    """
+    for b in BOOLEAN_ARGS:
+        if b in kwargs:
+            if kwargs[b]:
+                kwargs[b] = 'true'
+            else:
+                kwargs[b] = 'false'
+    for a in ARRAY_ARGS:
+        if a in kwargs:
+            value = kwargs[a]
+            if isinstance(value, Sequence):
+                kwargs[a] = json.dumps(list(value))
+            elif isinstance(value, basestring):
+                kwargs[a] = json.dumps(value)
+    for s in STRING_ARGS:
+        if s in kwargs:
+            kwargs[s] = json.dumps(kwargs[s])
+    return kwargs
+
+
+class Index(object):
+    """
+    _Index_
+
+    Sliceable and iterable interface to CouchDB View like things.
+
+    Instantiated with the raw data callable such as the
+    CloudantDatabase.all_docs or View.__call__ reference used to
+    retrieve data, the index can also store optional extra agrs for
+    customisation and supports efficient, paged iteration over the
+    results to avoid large views blowing up memory
+
+    """
+    def __init__(self, method_ref, **options):
+        self.options = options
+        self._ref = method_ref
+        self._page_size = options.pop("page_size", 100)
+        self._valid_args = ALL_ARGS
+
+    def _prepare_extras(self):
+        """
+        check that extra params are expected/valid
+        """
+        for k in self.options:
+            if k not in self._valid_args:
+                raise ValueError("Invalid argument: {0}".format(k))
+        extras = python_to_couch(self.options)
+        return extras
+
+    def __getitem__(self, key):
+        """
+        implement key access and slicing
+
+        """
+        extras = self._prepare_extras()
+        if isinstance(key, basestring):
+            data = self._ref(key=key, **extras)
+            return data['rows']
+
+        if isinstance(key, list):
+            data = self._ref(key=key, **extras)
+            return data['rows']
+
+        if isinstance(key, slice):
+            # slice is startkey and endkey if str or array
+            str_or_none_start = isinstance(
```

Comment:
going to keep this expanded for a while so that it is easy to tweak behaviour. Want to cover all the edge cases etc.